File: Version-JavaScript/models/Product.js
import { v4 as uuidv4 } from 'uuid';

//! Implementation of class and methods for CRUD operations (Create, Read, Update, Delete) for managing products in the system
class Product {
    constructor(name, price, quantity, minQuantity = 0, category) {
        if (!name || !price || !quantity || !category) {
            throw new Error("All required fields must be provided!");
        }

        this.id = uuidv4();
        this.name = name;
        this.price = price;
        this.quantity = quantity;
        this.minQuantity = minQuantity;
        this.category = category;
    }

	//! Method to Add a new product to the inventory.
	createProduct(supermaket) {
		supermaket.product.push(this);
	
		console.log(`Product successful added to the supermaket "${supermaket.name}" !`);
	};
	
	// Method Retrieve details of a product
	getProductInfos() {
		console.log(`
		Product Infos:
		-----------------
		
		ID: ${this.id}
		Name: ${this.name}
		Price: ${this.price} Francs
		Quantity: ${this.quantity}
		Minimal Quantity: ${this.minQuantity}
		Category: ${this.category}
		`);
	};
	
	//! Method to Update the details (price or quantity) of a product. 
	updateProduct(newName, newPrice, newQuantity, newCategory) { // jsut one argument is necessary, the rest is optional here...
        if(!newName && !newPrice && !newQuantity && !newCategory)
        	throw new Error("y've to enter at less one information !!!");
        	
        if(newName) this.name = newName;
    	if(newPrice) this.price = newPrice
    	if(newQuantity) this.quantity = newQuantity
    	if(newCategory) this.category = newCategory
    	
    	console.log("Success Information !");
    }

	//! Method to Remove a product from the inventory.
	deleteProduct(supermaket) {
		supermaket.products = supermaket.products.filter(product => product.id !== this.id);
		
		console.log(`This product has just successfully been deleted!`);
	};
	
    restock(quantity) {
        this.quantity += quantity;
        console.log(`${quantity} units added. New stock: ${this.quantity}`);
    };
    
    //! Method to Check if a product is in stock
	checkInventory() {
		if(this.quantity > this.minQuantity) {
			console.log(`Product is on stock !`);
			return true;
		} else {
			console.log(`Product is not on stock !`);
			return false;
		}
	};
}

export default Product;

// Test
const product = new Product("Savon", 400, 100, 10, "Domestique");
// product.getProductInfos();

File: Version-Python/models/product.py
from uuid import uuid4

class Product:
    def __init__(self, name, price, quantity, min_quantity=0, category=None):
        if not all([name, price, quantity, category]):
            raise ValueError("All required fields must be provided!")

        self.id = str(uuid4())
        self.name = name
        self.price = price
        self.quantity = quantity
        self.min_quantity = min_quantity
        self.category = category

    def create_product(self, supermarket):
        supermarket.products.append(self)
        print(f'Product successful added to the supermarket "{supermarket.name}"!')

    def get_product_infos(self):
        print(f"""
        Product Infos:
        -----------------
        
        ID: {self.id}
        Name: {self.name}
        Price: {self.price} Francs
        Quantity: {self.quantity}
        Minimal Quantity: {self.min_quantity}
        Category: {self.category}
        """)

    def update_product(self, new_name=None, new_price=None, new_quantity=None, new_category=None):
        if not any([new_name, new_price, new_quantity, new_category]):
            raise ValueError("You have to enter at least one information!")
            
        if new_name:
            self.name = new_name
        if new_price:
            self.price = new_price
        if new_quantity:
            self.quantity = new_quantity
        if new_category:
            self.category = new_category
            
        print("Success Information!")

    def delete_product(self, supermarket):
        supermarket.products = [p for p in supermarket.products if p.id != self.id]
        print("This product has just successfully been deleted!")

    def restock(self, quantity):
        self.quantity += quantity
        print(f"{quantity} units added. New stock: {self.quantity}")

    def check_inventory(self):
        if self.quantity > self.min_quantity:
            print("Product is on stock!")
            return True
        else:
            print("Product is not on stock!")
            return False
